---
title: "Clustering & ME: TCGA"
output: html_document
---
### Load datasets using CuratedPCa and add fusion info to TCGA
Load TCGA mae
```{r, message = FALSE}
library(curatedPCaData)
TCGA <- curatedPCaData::mae_tcga
```
Add fusion status to TCGA mae
```{r}
library(readxl)
TCGA.annotations <- read_xlsx("TCGA sample annotations.xlsx")
for (i in 1:333) {
  index <- which(TCGA@colData@listData$sample_name[i] == TCGA.annotations$SAMPLE_ID)
  TCGA@colData@listData$ERG_fusion_CNA[i] <- TCGA.annotations$Subtype[index]
}
```
### Subset TCGA by fusion status and compute enrichment
Subset CNA profiles by fusion status
```{r}
# TCGA subset
TCGA.CNA <- as.data.frame(TCGA@ExperimentList@listData$cna.gistic)
TCGA.CNA.fusion.pos <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "1-ERG" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "2-ETV1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "3-ETV4" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="4-FLI1")]
TCGA.CNA.fusion.neg <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "5-SPOP" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "6-FOXA1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "7-IDH1" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="8-other")]
```
Format CNA profiles for enrichment analysis
```{r}
# Format TCGA
TCGA.CNA.fusion.pos <- cbind(rownames(TCGA.CNA), TCGA.CNA.fusion.pos)
colnames(TCGA.CNA.fusion.pos)[1] <- "Hugo_Symbol"
TCGA.CNA.fusion.neg <- cbind(rownames(TCGA.CNA), TCGA.CNA.fusion.neg)
colnames(TCGA.CNA.fusion.neg)[1] <- "Hugo_Symbol"
```
Compute CNA enrichment in context of ETS fusions
```{r, message=FALSE}
source("analysis scripts.R")
TCGA.loss.enrichment <- concurrent.loss.data(TCGA.CNA.fusion.neg,TCGA.CNA.fusion.pos)
```

### Visualize genomic loss for threshold
```{r, message=FALSE}
#Set threshold for downstream analysis
cutoff <- 0.1
#Generate dataframe for plotting threshold genes
mp <- (TCGA.loss.enrichment[TCGA.loss.enrichment$`loss p val` <= cutoff,])[,c(1,9)]
#Plot threshold genes on manhattan plot
source("manhattan.R")
mp <- manhattan.object(mp)
kp <- kpPlotManhattan(karyoplot = plotKaryotype(plot.type = 1), 
                      data = mp,
                      points.col = "Navy",
                      genomewide.col = "white",
                      suggestive.col = "white")
```
```{r}
```
### Determine loss regions harbored by each patient
Filter out genes under the threshold for TCGA fusion neg patients
```{r}
source("annotate.R")
index <- TCGA.loss.enrichment$`loss p val` <= cutoff
threshold.genes <- annotate(TCGA.loss.enrichment[index,c(1,9)])
input <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "5-SPOP" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "6-FOXA1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "7-IDH1" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="8-other")]
index <- c()
for (i in 1:nrow(input)){
  if (sum(rownames(input)[i] == threshold.genes$Hugo_Symbol) == 0)
  {index[i] = FALSE}
  if (sum(rownames(input)[i] == threshold.genes$Hugo_Symbol) == 1)
  {index[i] = TRUE}
}
filtered.threshold <- input[index,]
```
Create binary matrix of patients by CNA < 0
``` {r}
binary <- filtered.threshold
binary[binary >= 0] <- 0
binary[binary < 0] <- 1
```
Annotate genes in binary matrix with gene ID, chromosome, band, and locus
```{r, warning=FALSE}
main <- data.frame(rownames(binary))
# Annotate Entrez Gene ID
mapped_HGNC <- as.data.frame(org.Hs.egSYMBOL[mappedkeys(org.Hs.egSYMBOL)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main[i,1] == mapped_HGNC$symbol))
  if (length(index) == 1)
  {main[i,2] <- mapped_HGNC$gene_id[index]}
  else {main[i,2] <- NA}
  }
colnames(main) <- c("Hugo_Symbol", "Gene ID")
main <- na.omit(main)
# Annotate Chromosome
mapped_CHR <- as.data.frame(org.Hs.egCHR [mappedkeys(org.Hs.egCHR)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHR$gene_id))
  if (length(index) == 1)
  {main[i,3] <- mapped_CHR$chromosome[index][1]}
  else {main[i,3] <- NA}
}
colnames(main)[3] <- "Chr"
# Annotate Band
mapped_CHRBAND <- as.data.frame(org.Hs.egMAP[mappedkeys(org.Hs.egMAP)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHRBAND$gene_id))
  if (length(index) == 1)
  {main[i,4] <- mapped_CHRBAND$cytogenetic_location[index][1]}
  else {main[i,4] <- NA}
  }
colnames(main)[4] <- "Band"
# Annotate by Chr-Arm
for (i in 1:nrow(main)){
  # Find index of "p" or "q" character in the Band column 
  if (unlist(gregexpr('q', main[i,4])) > 0)
  {index <- unlist(gregexpr('q', main[i,4]))}
  if (unlist(gregexpr('p', main[i,4])) > 0)
  {index <- unlist(gregexpr('p', main[i,4]))}
  # Input Char-Arm annotation
  main[i,5] <- substr(main[i,4], 1, index)
}
colnames(main)[5] <- "Locus"
```
Summary of copy number loss genes meeting threshold
```{r}
head(main)
tail(main)
```
Create binary matrix of patients by loci alteration < 0
```{r}
# Create reference data frame
binary.loci <- as.data.frame(cbind(main$Locus, binary))
colnames(binary.loci)[1] <- "Locus"
# Initialize new data frame to store output
output <- data.frame()
# Create new data frame with rows representing locus
  # Summary of lost loci
loci <- binary.loci %>% count(Locus)
  # Determine if patients harbor loss of a given loci. Compute each loci and add to main
  # data frame row-wise to build final table for ME analysis.
for (i in 1:nrow(loci)){
  row <- c()
  locus <- loci[i,1]
  locus.loss <- binary.loci[binary.loci[,1] == locus,]
  # Determine if each patient has loss of at least one gene from the locus. If they do,
  # give patient a 1 for loss. If not, give the patient a zero.
  for (k in 2:ncol(locus.loss)){
    if (sum((locus.loss[,k]) == "1") > 0)
    {row[k-1] <- 1} 
    if (sum((locus.loss[,k]) == "0") > 0)
    {row[k-1] <- 0} 
  }
  output[i,1:length(row)] <- row
  rownames(output)[i] <- locus
}
colnames(output) <- colnames(locus.loss)[2:ncol(locus.loss)]
head(output)
```
Verify number of patients harboring lost of at least one of the genes meeting cutoff
```{r}
coverage <- 0
for (i in (1:ncol(output))){
  if (sum(output[,i]) > 0)
  {coverage <- coverage + 1}
}
print(coverage)
```
The number of ETS- TCGA patients with loss of at least one gene from one of these loci is `r coverage`, and the total number of ETS- TCGA patients is `r ncol(TCGA.CNA.fusion.neg)-1`.
```{r}```
### Loci alteration profiles
```{r}
x <- as.matrix(output)
heatmap(x, Rowv = NA, labCol = NA, xlab = "ETS- Patients", ylab = "Chromosome Arm",
        col = c("Grey", "Red"), scale = "col", cexRow = 0.7)
#legend(x = -.07, y = 0.5, legend <- c("Loss", "Unaltered"),
       #fill = c("red", "grey"), cex = 1, box.lwd = 0)
```
```{r}```
### Mutual exclusivity
```{r, message=FALSE}
library(Rediscover)
library(pheatmap)
library(RColorBrewer)
#Calculate probability of copy number loss
output <- as.matrix(output)
p.loss <- getPM(output)
z1 <- as.matrix(p.loss)
#Calculate mutual exclusivity of copy number loss
ME <- getMutex(output, p.loss)
#Heat plot
z2 <- as.matrix(ME)
z2 <- -log10(z2)
rownames(z2) <- rownames(output)
colnames(z2) <- rownames(output)
tri <- lower.tri(z2, diag = FALSE)
z3 <- z2
z3[tri] <- NA
z3 <- z3[order(nrow(z3):1),]
pheatmap(z3, color = brewer.pal(9,"Blues"),
         main ="Mutual Exclusivity\nlog10(p-val)",
         display_numbers = TRUE, cluster_rows = FALSE, cluster_cols = FALSE, 
         na_col = "white")
```

