---
title: "Clustering and ME for TCGA"
output: html_document
---
### Load datasets add fusion info to TCGA
Load mae objects, take complete cases
```{r, message = FALSE}
library(curatedPCaData)
TCGA <- curatedPCaData::mae_tcga
TCGA <- intersectColumns(TCGA)
```
### Subset TCGA dataset by ETS-family fusion status and compute copy number loss enrichment
Subset CNA profiles by fusion status
``` {r}
#TCGA subset
TCGA.fusion.neg <- TCGA[,TCGA$ERG_fusion_GEX==0]
TCGA.fusion.pos <- TCGA[,TCGA$ERG_fusion_GEX==1]
```
Format CNA profiles for enrichment analysis
```{r}
#Format TCGA
TCGA.fusion.pos <- as.data.frame(cbind(rownames(TCGA.fusion.pos@ExperimentList@listData$cna.gistic),
                         TCGA.fusion.pos@ExperimentList@listData$cna.gistic))
colnames(TCGA.fusion.pos)[1] <- "Hugo_Symbol"
TCGA.fusion.neg <- as.data.frame(cbind(rownames(TCGA.fusion.neg@ExperimentList@listData$cna.gistic),
                         TCGA.fusion.neg@ExperimentList@listData$cna.gistic))
colnames(TCGA.fusion.neg)[1] <- "Hugo_Symbol"
```
Compute CNA enrichment in context of ETS fusions
```{r, message=FALSE}
source("analysis scripts.R")
TCGA.loss.enrichment <- concurrent.loss.data(TCGA.fusion.neg, TCGA.fusion.pos)
```
### Visualize ETS- genomic loss subject to threshold
Filter out genes lost in ETS+ patients
```{r, warning=FALSE}
x <- c()
ERGneg.genes <- TCGA.loss.enrichment
for (i in 1:nrow(ERGneg.genes)){
  if (ERGneg.genes$`Subset loss ratio`[i] >= ERGneg.genes$`Dataset loss ratio`[i])
  {x[i] = TRUE} else {x[i] = FALSE}
}
ERGneg.genes <- ERGneg.genes[x,]
```
Further filter out genes that are (1) under the cutoff and (2) lacking annotation.
```{r}
cutoff <- 0.01
ERGneg.genes <- annotate(ERGneg.genes[ERGneg.genes$`loss p val` <= cutoff,c(1,9)])
```
Visualize genomic loss in ETS- patients subject to threshold
```{r, message=FALSE}
#Create manhattan object
source("manhattan.R")
mp <- manhattan.object(ERGneg.genes[,c(1,2)])
#Plot manhattan object
kp <- kpPlotManhattan(karyoplot = plotKaryotype(plot.type = 2), 
                      data = mp,
                      points.col = "Navy",
                      genomewide.col = "white",
                      suggestive.col = "white")
```
```{r}
```
### Determine statistically significant loss regions harbored by ERG- patients
Filter out ganees from ETS- CNA profiles that are (1) not lost in ETS- patients, (2) do not meet cutoff, and (3) lack annotation
```{r}
x <- c()
for (i in 1:nrow(TCGA.fusion.neg)){
  if (sum(TCGA.fusion.neg[i,1] == ERGneg.genes$Hugo_Symbol) > 0)
  {x[i] <- TRUE} else {x[i] <- FALSE}
}
TCGA.fusion.neg.filtered <- TCGA.fusion.neg[x,]
```
Create binary matrix of CNA < 0 for ETS- patients
```{r}
binary <- TCGA.fusion.neg.filtered
binary[binary >= 0] <- 0
binary[binary < 0] <- 1
```
Annotate genes in binary matrix with gene ID, chromosome, band, and locus
``` {r, warning=FALSE}
main <- data.frame(rownames(binary))
# Annotate Entrez Gene ID
mapped_HGNC <- as.data.frame(org.Hs.egSYMBOL[mappedkeys(org.Hs.egSYMBOL)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main[i,1] == mapped_HGNC$symbol))
  if (length(index) == 1)
  {main[i,2] <- mapped_HGNC$gene_id[index]}
  else {main[i,2] <- NA}
  }
colnames(main) <- c("Hugo_Symbol", "Gene ID")
main <- na.omit(main)
# Annotate Chromosome
mapped_CHR <- as.data.frame(org.Hs.egCHR [mappedkeys(org.Hs.egCHR)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHR$gene_id))
  if (length(index) == 1)
  {main[i,3] <- mapped_CHR$chromosome[index][1]}
  else {main[i,3] <- NA}
}
colnames(main)[3] <- "Chr"
# Annotate Band
mapped_CHRBAND <- as.data.frame(org.Hs.egMAP[mappedkeys(org.Hs.egMAP)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHRBAND$gene_id))
  if (length(index) == 1)
  {main[i,4] <- mapped_CHRBAND$cytogenetic_location[index][1]}
  else {main[i,4] <- NA}
  }
colnames(main)[4] <- "Band"
# Annotate by Chr-Arm
for (i in 1:nrow(main)){
  # Find index of "p" or "q" character in the Band column 
  if (unlist(gregexpr('q', main[i,4])) > 0)
  {index <- unlist(gregexpr('q', main[i,4]))}
  if (unlist(gregexpr('p', main[i,4])) > 0)
  {index <- unlist(gregexpr('p', main[i,4]))}
  # Input Char-Arm annotation
  main[i,5] <- substr(main[i,4], 1, index)
}
colnames(main)[5] <- "Locus"
```
Summary of genes loss-enriched in ETS- patients
```{r, message = FALSE}
library(DT)
main <- main[order(main$Chr),]
rownames(main) <- 1:nrow(main)
datatable(main)
```
Create binary matrix of patients by loci alteration < 0
```{r}
# Create reference data frame
binary.loci <- as.data.frame(cbind(main$Locus, binary))
colnames(binary.loci)[1] <- "Locus"
# Initialize new data frame to store output
output <- data.frame()
# Create new data frame with rows representing locus
  # Summary of lost loci
loci <- binary.loci %>% count(Locus)
  # Determine if patients harbor loss of a given loci. Compute each loci and add to main
  # data frame row-wise to build final table for ME analysis.
for (i in 1:nrow(loci)){
  row <- c()
  locus <- loci[i,1]
  locus.loss <- binary.loci[binary.loci[,1] == locus,]
  # Determine if each patient has loss of at least one gene from the locus. If they do,
  # give patient a 1 for loss. If not, give the patient a zero.
  for (k in 2:ncol(locus.loss)){
    if (sum((locus.loss[,k]) == "1") > 0)
    {row[k-1] <- 1} else {row[k-1] <- 0}
  }
  output[i,1:length(row)] <- row
  rownames(output)[i] <- locus
}
colnames(output) <- colnames(locus.loss)[2:ncol(locus.loss)]
datatable(output)
```
Verify number of patients harboring loss of at least one of the genes meeting cutoff
```{r}
output <- as.matrix(output[,2:ncol(output)])
coverage <- 0
for (i in (1:ncol(output))){
  if (sum(output[,i]) > 0)
  {coverage <- coverage + 1}
}
print(coverage)
print(ncol(TCGA.fusion.neg)-1)
```
The number of ETS- TCGA patients with loss of at least one gene from one of these loci is `r coverage`, and the total number of ETS- TCGA patients is `r ncol(TCGA.fusion.neg)-1`. Thus, the ratio of ETS- patients covered at this threshold is `r coverage / (ncol(TCGA.fusion.neg)-1)`. This coverage should match the ratio of ETS- patients reported in the Determinring Threshold.Rmd file.
### Loci alteration profiles
```{r, message=FALSE}
library(pheatmap)
heatmap(output, scale = "none", col = c("grey", "red"), Colv = NA, labCol = NA,
        ylab = "Chromosome Arm", xlab = "ETS- patients")
legend(x = 0.875, y = 0.2, legend = c("Deletion", "Unaltered"), fill = c("red", "grey"), box.lwd = 0)
```
```{r}```
### Mutual exclusivity
```{r, message=FALSE}
library(Rediscover)
library(RColorBrewer)
#Calculate probability of copy number loss
p.loss <- getPM(output)
#Calculate mutual exclusivity of copy number loss
ME <- getMutex(output, p.loss)
#Heat plot
z2 <- as.matrix(ME)
rownames(z2) <- rownames(output)
colnames(z2) <- rownames(output)
tri <- lower.tri(z2, diag = FALSE)
z3 <- z2
z3[tri] <- NA
z3 <- z3[order(nrow(z3):1),]
pheatmap(z3, color = brewer.pal(5, "Reds"), cluster_rows = FALSE, cluster_cols = FALSE, main = "Mutual Exclusivity\n(p-val)")
```

