---
title: "Exclusivity Analysis"
output: html_document
---
### Load datasets using CuratedPCa and add fusion info to TCGA
Load mae objects
```{r, message = FALSE}
library(curatedPCaData)
TCGA <- curatedPCaData::mae_tcga
Abida <- curatedPCaData::mae_abida
```
Add fusion status to TCGA mae
```{r}
library(readxl)
TCGA.annotations <- read_xlsx("TCGA sample annotations.xlsx")
for (i in 1:333) {
  index <- which(TCGA@colData@listData$sample_name[i] == TCGA.annotations$SAMPLE_ID)
  TCGA@colData@listData$ERG_fusion_CNA[i] <- TCGA.annotations$Subtype[index]
}
```
### Subset TCGA and Abida by fusion status and compute enrichment
Subset CNA profiles by fusion status
```{r}
#TCGA subset
TCGA.CNA <- as.data.frame(TCGA@ExperimentList@listData$cna.gistic)
TCGA.CNA.fusion.pos <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "1-ERG" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "2-ETV1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "3-ETV4" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="4-FLI1")]
TCGA.CNA.fusion.neg <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "5-SPOP" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "6-FOXA1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "7-IDH1" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="8-other")]
#Abida subset
Abida.CNA <- as.data.frame(Abida@ExperimentList@listData$cna.gistic)
Abida.CNA.fusion.pos <- Abida.CNA[,Abida@colData@listData$ERG_fusion_GEX==1]
Abida.CNA.fusion.neg <- Abida.CNA[,Abida@colData@listData$ERG_fusion_GEX==0]
```
Format CNA profiles for enrichment analysis
```{r}
#Format TCGA
TCGA.CNA.fusion.pos <- cbind(rownames(TCGA.CNA), TCGA.CNA.fusion.pos)
colnames(TCGA.CNA.fusion.pos)[1] <- "Hugo_Symbol"
TCGA.CNA.fusion.neg <- cbind(rownames(TCGA.CNA), TCGA.CNA.fusion.neg)
colnames(TCGA.CNA.fusion.neg)[1] <- "Hugo_Symbol"
#Format Abida
Abida.CNA.fusion.pos <- cbind(rownames(Abida.CNA), Abida.CNA.fusion.pos)
colnames(Abida.CNA.fusion.pos)[1] <- "Hugo_Symbol"
Abida.CNA.fusion.neg <- cbind(rownames(Abida.CNA), Abida.CNA.fusion.neg)
colnames(Abida.CNA.fusion.neg)[1] <- "Hugo_Symbol"
```
Compute CNA enrichment in context of ETS fusions
```{r, message=FALSE}
source("analysis scripts.R")
TCGA.loss.enrichment <- concurrent.loss.data(TCGA.CNA.fusion.neg,TCGA.CNA.fusion.pos)
#Abida.loss.enrichment <-concurrent.loss.data(Abida.CNA.fusion.neg,Abida.CNA.fusion.pos)
```
### Determine threshold for identifying chromosomal loss regions
Determine threshold for genes
```{r, message=FALSE}
#Set threshold index
threshold <- TCGA.loss.enrichment$`loss p val` < .1
#Determine genes within threshold
genes <- TCGA.CNA.fusion.neg[threshold,]
#Determine the number of patients harboring CNA of thresholded genes
x <- c(TRUE)
for (i in 2:ncol(TCGA.CNA.fusion.neg))
  {CNA.sum <- sum(abs(genes[,i]))
  if (CNA.sum == 0)
    {x[i] = FALSE}
  if (CNA.sum > 0)
    {x[i] = TRUE}
}
patients <- TCGA.CNA.fusion.neg[,x]
```
Threshold set at p-val < 0.2. The number of ERG- TCGA patients with alteration of genes meeting this threshold is `r ncol(patients)-1`, and the total number of ERG- TCGA patients is `r ncol(TCGA.CNA.fusion.neg)-1`.
Show loci harboring significant copy number loss
```{r, message=FALSE}
#Generate dataframe for plotting threshold genes
loss.genes <- TCGA.loss.enrichment[TCGA.loss.enrichment$`loss p val` < 0.2,]
loss.genes <- loss.genes[,c(1,9)]
#Plot threshold genes on manhattan plot. This visualization shows statistically enriched regions of copy number loss in ERG- TCGA patients.
source("manhattan.R")
mp <- manhattan.object(loss.genes)
kp <- kpPlotManhattan(karyoplot = plotKaryotype(plot.type = 1), 
                      data = mp,
                      genomewide.col = "white",
                      suggestive.col = "white")
```
```{r}
```
### Enumerate regions of chromosomal loss
Highlight regions of chromosomal loss for mutual exclusivity analysis
```{r, message=FALSE}
#Cytogenetic band locations of threshold genes
source("annotate.R")
loss.regions <- annotate(loss.genes)
summary.loss.regions <- loss.regions %>% count(Band)
#Highlight regions of chromosomal loss that encompass threshold genes. 
#Used IGV to determine BP ranges spanning the highlighted band locations.
#Chromosome 1 loss occurring from 1q31-1q33, and 1p12-1p21
Chr1p <- "chr1:90,000,000-120,000,000"
Chr1q <- "chr1:186,000,000-214,000,000"
#Chromosome 2 loss occurring from 2q32-2q37
Chr2 <- "chr2:190,000,000-240,000,000"
#Chromosome 3 loss occurring from 3q13-3q21
Chr3 <- "chr3:100,000,000-135,000,000"
#Chromosome 5 loss occurring from 5q21-5q32
Chr5 <- "chr5:100,000,000-152,000,000"
#Chromosome 6 loss occurring from 6p22-6p25 and 6q13-6q27
Chr6p <- "chr6:1-30,000,000"
Chr6q <- "chr6:100,000,000-171,000,000"
#Plot highlighted loss regions
kp <- kpPlotManhattan(karyoplot = plotKaryotype(plot.type = 1), 
                      data = mp, 
                      highlight = c(Chr1p, Chr1q, Chr2, Chr3, Chr5, Chr6p, Chr6q),
                      highlight.col = "blue",
                      suggestive.col = "white",
                      genomewide.col = "white")
```
```{r}```
### Determine regions of loss harbroed by each patient
Filter out genes under the threshold for TCGA fusion neg patients
```{r}
genes <- loss.regions[,c(1,4)]
genes[,2] <- genes[,2]
input <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "5-SPOP" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "6-FOXA1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "7-IDH1" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="8-other")]
index <- c()
for (i in 1:nrow(input)){
  if (sum(rownames(input)[i] == genes$Hugo_Symbol) == 0)
  {index[i] = FALSE}
  if (sum(rownames(input)[i] == genes$Hugo_Symbol) == 1)
  {index[i] = TRUE}
}
filtered.threshold <- input[index,]
```
Filter out genes that are not highlighted
```{r}
index <- c()
#Create character vectors holding cytoband names of highlighted regions
Chr1p <- as.character(summary.loss.regions$Band[102:108])
Chr1q <- as.character(summary.loss.regions$Band[111:115])
Chr2 <- as.character(summary.loss.regions$Band[131:143])
Chr3 <- as.character(summary.loss.regions$Band[c(35,144:148)])
Chr5 <- as.character(summary.loss.regions$Band[162:173])
Chr6p <- as.character(summary.loss.regions$Band[179:185])
Chr6q <- as.character(summary.loss.regions$Band[178:185])
for (i in 1:nrow(loss.regions)){
  #Index highlighted genes on Chr1p
  if ((sum(loss.regions$Band[i] == Chr1p)) > 0)
  {index[i] = TRUE}
  #Index highlighted genes on Chr1q
  if ((sum(loss.regions$Band[i] == Chr1q)) > 0)
  {index[i] = TRUE}
  #Index highlighted genes on Chr2
  if ((sum(loss.regions$Band[i] == Chr2)) > 0)
  {index[i] = TRUE}
  #Index highlighted genes on Chr3
  if ((sum(loss.regions$Band[i] == Chr3)) > 0)
  {index[i] = TRUE}
  #Index highlighted genes on Chr5
  if ((sum(loss.regions$Band[i] == Chr5)) > 0)
  {index[i] = TRUE}
  #Index highlighted genes on Chr6p
  if ((sum(loss.regions$Band[i] == Chr6p)) > 0)
  {index[i] = TRUE}
  #Index highlighted genes on Chr6q
  if ((sum(loss.regions$Band[i] == Chr6q)) > 0)
  {index[i] = TRUE}
  }
filtered.highlight <- na.omit(filtered.threshold[index,])
```
Create binary matrix containing CNA events < 0
```{r}
binary <- filtered.highlight
binary[binary > 0] <- 0
binary[binary < 0] <- -1
binary <- as.matrix(abs(binary))
```
Annotate genes in binary matrix with bands and regions of chromosomal loss
```{r}
main <- data.frame(rownames(binary))
#Annotate Entrez Gene ID
#Obtain dataframe of Gene ID mapped to HGNC Symbol
mapped_HGNC <- as.data.frame(org.Hs.egSYMBOL[mappedkeys(org.Hs.egSYMBOL)])
#Map Entrez Gene ID to HGNC Symbols in main dataframe
for (i in 1:nrow(main)){
  index = as.numeric(which(main[i,1] == mapped_HGNC$symbol))
  if (length(index) == 1)
  {main[i,2] <- mapped_HGNC$gene_id[index]}
  else {main[i,2] <- NA}
  }
colnames(main) <- c("Hugo_Symbol", "Gene ID")
main <- na.omit(main)
#Annotate Band
mapped_CHRBAND <- as.data.frame(org.Hs.egMAP[mappedkeys(org.Hs.egMAP)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHRBAND$gene_id))
  if (length(index) == 1)
  {main[i,3] <- mapped_CHRBAND$cytogenetic_location[index][1]}
  else {main[i,3] <- NA}
  }
colnames(main)[3] <- "Band"
#Annotate genes with loss region
for (i in 1:nrow(main)){
  if (sum(main[i,3] == Chr1p) > 0)
  {main[i,4] <- "Chr1p"}
  if (sum(main[i,3] == Chr1q) > 0)
  {main[i,4] <- "Chr1q"}
  if (sum(main[i,3] == Chr2) > 0)
  {main[i,4] <- "Chr2"}
  if (sum(main[i,3] == Chr3) > 0)
  {main[i,4] <- "Chr3"}
  if (sum(main[i,3] == Chr5) > 0)
  {main[i,4] <- "Chr5"}
  if (sum(main[i,3] == Chr6p) > 0)
  {main[i,4] <- "Chr6p"}
  if (sum(main[i,3] == Chr6q) > 0)
  {main[i,4] <- "Chr6q"}
}
colnames(main)[4] <- "loss region"
```
Annotate binary matrix with loss region
```{r}
binary.lossregion <- cbind.data.frame(main$`loss region`, binary)
binary.lossregion <- na.omit(binary.lossregion)
colnames(binary.lossregion)[1] <- "loss region"
```
Create binary matrix of chromosomal loss by patient
```{r}
#Compute Chr1q loss across patients
Chr1q <- data.frame()
for (i in 2:ncol(binary.lossregion)){
  region <- binary.lossregion[binary.lossregion$`loss region`=="Chr1q",i]
  if (sum(region) > 0)
  {Chr1q[1,i-1] <- 1} else {Chr1q[1,i-1] <- 0}
  colnames(Chr1q)[i-1] <- colnames(binary.lossregion)[i]
  rownames(Chr1q) <- "Chr1q"
}
#Compute Chr1q loss across patients
Chr1p <- data.frame()
for (i in 2:ncol(binary.lossregion)){
  region <- binary.lossregion[binary.lossregion$`loss region`=="Chr1p",i]
  if (sum(region) > 0)
  {Chr1p[1,i-1] <- 1} else {Chr1p[1,i-1] <- 0}
  colnames(Chr1p)[i-1] <- colnames(binary.lossregion)[i]
  rownames(Chr1p) <- "Chr1p"
}
#Compute Chr2 loss across patients
Chr2 <- data.frame()
for (i in 2:ncol(binary.lossregion)){
  region <- binary.lossregion[binary.lossregion$`loss region`=="Chr2",i]
  if (sum(region) > 0)
  {Chr2[1,i-1] <- 1} else {Chr2[1,i-1] <- 0}
  colnames(Chr2)[i-1] <- colnames(binary.lossregion)[i]
  rownames(Chr2) <- "Chr2"
}
#Compute Chr3 loss across patients
Chr3 <- data.frame()
for (i in 2:ncol(binary.lossregion)){
  region <- binary.lossregion[binary.lossregion$`loss region`=="Chr3",i]
  if (sum(region) > 0)
  {Chr3[1,i-1] <- 1} else {Chr3[1,i-1] <- 0}
  colnames(Chr3)[i-1] <- colnames(binary.lossregion)[i]
  rownames(Chr3) <- "Chr3"
}
#Compute Chr5 loss across patients
Chr5 <- data.frame()
for (i in 2:ncol(binary.lossregion)){
  region <- binary.lossregion[binary.lossregion$`loss region`=="Chr5",i]
  if (sum(region) > 0)
  {Chr5[1,i-1] <- 1} else {Chr5[1,i-1] <- 0}
  colnames(Chr5)[i-1] <- colnames(binary.lossregion)[i]
  rownames(Chr5) <- "Chr5"
}
#Computer Chr6p loss across patients
Chr6p <- data.frame()
for (i in 2:ncol(binary.lossregion)){
  region <- binary.lossregion[binary.lossregion$`loss region`=="Chr6p",i]
  if (sum(region) > 0)
  {Chr6p[1,i-1] <- 1} else {Chr6p[1,i-1] <- 0}
  colnames(Chr6p)[i-1] <- colnames(binary.lossregion)[i]
  rownames(Chr6p) <- "Chr6p"
}
#Computer Chr6q loss across patients
Chr6q <- data.frame()
for (i in 2:ncol(binary.lossregion)){
  region <- binary.lossregion[binary.lossregion$`loss region`=="Chr6q",i]
  if (sum(region) > 0)
  {Chr6q[1,i-1] <- 1} else {Chr6q[1,i-1] <- 0}
  colnames(Chr6q)[i-1] <- colnames(binary.lossregion)[i]
  rownames(Chr6q) <- "Chr6q"
}
regional.loss <- rbind.data.frame(Chr1q, Chr1p, Chr2, Chr3, Chr5, Chr6p, Chr6q)
```
Mutual Exclusivity Analysis
```{r, message=FALSE}
library(Rediscover)
library(pheatmap)
library(RColorBrewer)
#Calculate probability of mutation
regional.loss <- as.matrix(regional.loss)
p.loss <- getPM(regional.loss)
z1 <- as.matrix(p.loss)
#Calculate mutual exclusivity
ME <- getMutex(regional.loss, p.loss)
#Heat plot
z2 <- as.matrix(ME)
rownames(z2) <- rownames(regional.loss)
colnames(z2) <- rownames(regional.loss)
pheatmap(z2, color = brewer.pal(9,"Blues"),
         main ="Mutual Exclusivity (p-val)")
```
