---
title: "Figure 1 Analysis"
output: html_document
---
# Figure 1 Analysis
Load mae for each study
```{r}
TCGA <- curatedPCaData::mae_tcga
Abida <- curatedPCaData::mae_abida
```
Add fusion status to TCGA mae
```{r}
TCGA.annotations <- read_xlsx("TCGA sample annotations.xlsx")
for (i in 1:333) {
  index <- which(TCGA@colData@listData$sample_name[i] == TCGA.annotations$SAMPLE_ID)
  TCGA@colData@listData$ERG_fusion_CNA[i] <- TCGA.annotations$Subtype[index]
}
```
Subset CNA profiles by fusion status
```{r}
TCGA.CNA <- as.data.frame(TCGA@ExperimentList@listData$cna.gistic)
TCGA.CNA.fusion.pos <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "1-ERG" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "2-ETV1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "3-ETV4" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="4-FLI1")]
TCGA.CNA.fusion.neg <- TCGA.CNA[,(TCGA@colData@listData$ERG_fusion_CNA == "5-SPOP" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "6-FOXA1" |
                                    TCGA@colData@listData$ERG_fusion_CNA == "7-IDH1" |
                                    TCGA@colData@listData$ERG_fusion_CNA =="8-other")]
Abida.CNA <- as.data.frame(Abida@ExperimentList@listData$cna.gistic)
Abida.CNA.fusion.pos <- Abida.CNA[,Abida@colData@listData$ERG_fusion_GEX==1]
Abida.CNA.fusion.neg <- Abida.CNA[,Abida@colData@listData$ERG_fusion_GEX==0]
```
Format CNA profiles for enrichment analysis
```{r}
#TCGA
TCGA.CNA.fusion.pos <- cbind(rownames(TCGA.CNA), TCGA.CNA.fusion.pos)
colnames(TCGA.CNA.fusion.pos)[1] <- "Hugo_Symbol"
TCGA.CNA.fusion.neg <- cbind(rownames(TCGA.CNA), TCGA.CNA.fusion.neg)
colnames(TCGA.CNA.fusion.neg)[1] <- "Hugo_Symbol"
#Abida
Abida.CNA.fusion.pos <- cbind(rownames(Abida.CNA), Abida.CNA.fusion.pos)
colnames(Abida.CNA.fusion.pos)[1] <- "Hugo_Symbol"
Abida.CNA.fusion.neg <- cbind(rownames(Abida.CNA), Abida.CNA.fusion.neg)
colnames(Abida.CNA.fusion.neg)[1] <- "Hugo_Symbol"
```
Compute copy number loss enrichment
```{r}
source("analysis scripts.R")
TCGA.loss.enrichment <- concurrent.loss.data(TCGA.CNA.fusion.neg,TCGA.CNA.fusion.pos)
Abida.loss.enrichment <- concurrent.loss.data(Abida.CNA.fusion.neg,Abida.CNA.fusion.pos)
```
##Generalizable code for annotating genome-wide hit list
Create main data frame. Column 1 contains HGNC symbols, and Column 2 contains 
p-value from whatever type of computational analysis
```{r}
Hugo.symbol <- TCGA.loss.enrichment$Hugo_Symbol
p <- TCGA.loss.enrichment$`adjusted loss p val`
main <- data.frame(Hugo.symbol, p)
```
Annotate Entrez Gene ID
```{r}
#Obtain dataframe of Gene ID mapped to HGNC Symbol
mapped_HGNC <- as.data.frame(org.Hs.egSYMBOL[mappedkeys(org.Hs.egSYMBOL)])
#Map Entrez Gene ID to HGNC Symbols in main dataframe
for (i in 1:length(Hugo.symbol)){
  index = as.numeric(which(main$Hugo.symbol[i] == mapped_HGNC$symbol))
  if (length(index) == 1)
  {main[i,3] <- mapped_HGNC$gene_id[index]}
  else {main[i,3] <- NA}
}
colnames(main)[3] <- "Gene ID"
```
Annotate Chromosome Number
```{r}
#Obtain dataframe of Gene ID mapped to Chromosome Number
mapped_CHR <- as.data.frame(org.Hs.egCHR[mappedkeys(org.Hs.egCHR)])
#Change ChrX to 23 and ChrY to 24
for (i in 1:nrow(mapped_CHR)){
  if (mapped_CHR[i,2] == "X") {mapped_CHR[i,2] <- 23}
  if (mapped_CHR[i,2] == "Y") {mapped_CHR[i,2] <- 24}
}
#Map Entrez Gene ID to Chromosome Number in main dataframe
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHR$gene_id))
  if (length(index) == 1)
  {main[i,4] <- as.numeric(mapped_CHR$chromosome[index])}
  else {main[i,4] <- NA}
}
colnames(main)[4] <- "Chr"
```
Annotate Chromosome Location
```{r}
mapped_CHRLOC <- as.data.frame(org.Hs.egCHRLOC[mappedkeys(org.Hs.egCHRLOC)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHRLOC$gene_id))
  if (length(index) == 1)
  {main[i,5] <- abs(mapped_CHRLOC$start_location[index])}
  else {main[i,5] <- NA}
}
colnames(main)[5] <- "Location"
```
Annotate Cytogenetic Band
```{r}
mapped_CHRBAND <- as.data.frame(org.Hs.egMAP[mappedkeys(org.Hs.egMAP)])
for (i in 1:nrow(main)){
  index = as.numeric(which(main$`Gene ID`[i] == mapped_CHRBAND$gene_id))
  if (length(index) == 1)
  {main[i,6] <- mapped_CHRBAND$cytogenetic_location[index]}
  else {main[i,6] <- NA}
}
colnames(main)[6] <- "Band"
```
Clear NA from main dataframe
```{r}
main <- na.omit(main)
```
Annotate Position on Chromosome
```{r}
#Create a sorted vector with positional info for each gene on its' respective chromosome
genes.by.chr <- main[main$Chr==1,]
sorted <- sort(genes.by.chr$Loc)
#Assign position of each gene on its' respective chromosome for Chr 1
for (i in 1:nrow(genes.by.chr)){
  index <- which(sorted == genes.by.chr$Loc[i])[1]
  genes.by.chr[i,6] <- index
}
#Initialize manhattan plot data frame for appending chromosomes 2 - 24
mp <- genes.by.chr
#Assign position of each gene on its respective chromosome for Chr2-24
for (i in 2:24){
  genes.by.chr <- main[main$Chr==i,]
  sorted <- sort(genes.by.chr$Loc)
  #Assign position of each gene on its' respective chromosome
    for (k in 1:nrow(genes.by.chr)){
      index <- which(sorted == genes.by.chr$Loc[k])[1]
      genes.by.chr[k,6] <- index}
  #Append information for that chromosome to the mp data frame
  mp <- rbind(mp, genes.by.chr)
}
```
Format mp data frame for use with manhattan plot package
```{r}
mp <-  mp[,c(1, 2, 4, 6)]
colnames(mp) <- c("Gene", "p", "CHR", "BP")
mp <- na.omit(mp)
```
Make manhattan plot
```{r}
manhattan(mp, 
          chr = "CHR", 
          bp = "BP", 
          p = "p", 
          snp = "Gene")
```
